### ISSUE TITLE
Optimize annotation card rendering

### ISSUE THREAD

**Issue by robertknight** [hypothesis/client#1517]
Yesterday I was looking into why switching tabs on https://hypothes.is/docs/help felt a bit janky, mainly when switching to the "Annotations" tab. In the process I ended up sharing the steps to reproduce with the Preact developers and one of the core devs kindly did a deep dive and provided some useful insights which I'm capturing here.

Transcript (see comments from `developit` in particular):

<details>

```
robertknight Yesterday at 3:02 PM
What’s the current state of Preact profiling tools? I’m looking into an interaction that is sometimes laggy in our app (takes a few hundred ms, needs to be <100ms really) and an issue I’m noticing in both Safari and Chrome is that the very deep call stacks from recursive rendering are a bit of a PITA to work with in both Safari and Chrome’s profiling tools. The component/DOM tree is not trivial but not crazy deep either.
24 replies

cristian  18 hours ago
Marvin knows best. He added the repo to our org

cristian  18 hours ago
Curious, is your issue reproducing with React? If yes, then maybe use their devtool

robertknight  18 hours ago
Our app doesn’t use React at all. I don’t suspect Preact is at fault here, just that digging into what we’re doing that is causing the lag is frustrated by this particular issue.

mhagemeister  18 hours ago
I have prepared the code to collect timings etc but haven't worked on proper profiling integration yet (edited) 

mhagemeister  18 hours ago
It's mainly due to a lack of time on my part

developit  16 hours ago
@robertknight I wonder if there's a way you could send me over a CPU profile? I've been doing performance optimization stuff and am working through PR'ing a 2x speedup of tree diffing. (edited) 
:heart:
1


developit  16 hours ago
(I know it's not technically helping with the profile bit but it may address the underlying performance issue)

robertknight  15 hours ago
If you wanted to reproduce for yourself, you can go to a test page we have at https://hypothes.is/docs/help and click back and forth between the “Annotations” and “Page Notes” tabs at the top of the pane on the right-hand side of the page.
The app is 2/3rds of the way through an Angular v1 => Preact migration so there is a bunch of non-Preact stuff in the profiles. Most of the content in that iframe is by now rendered by Preact.
At this point we haven’t done any optimization with shouldComponentUpdate etc. I’m just starting to look at profiling to tune the experience.

robertknight  15 hours ago
These are the tabs I’m referring to.
annotation-tabs.png 
annotation-tabs.png



developit  11 hours ago
this is a pretty neat app

developit  11 hours ago
ok so quick profile shows this: the Excerpt  component has a useLayoutEffect hook that checks the clientHeight of a ref. It seems like that callback is being invoked a lot, and it happens right at the bottom of the stack. useCallback fires during rendering, which means layout is being constantly invalidated, then queried (back and forth), which causes really really expensive forced synchronous layout. (edited) 

developit  11 hours ago
This happens because there are multiple render roots being used, and althrough useLayoutEffect callbacks are queued up during rendering, they all fire synchronously as preact exits render() - when there are N render() calls, you get N layouts.

developit  11 hours ago
Also sentry installs a hook into Element.prototype.addEventListener that does some really inefficient debouncing.

developit  11 hours ago
Looking at the Markdown/Math rendering, DOMPurify is being used in a way that creates new HTML document contexts for every call to renderMathAndMarkdown().  This one is a bit trickier, but there are some options:
One of the more interesting ones would be to blast the unsanitized HTML into the DOM using dangerouslySetInnerHTML, but then DOMPurify it in-place (DOMPurify.sanitize(node,{IN_PLACE:true})). This doesn't work for sanitizing things like <script>, but it does for sanitizing attributes and whatnot.
Cache sanitized HTML. Caching locally at the component would avoid hitting DOMPurify on every re-render, but if you're going to cache you might as well do it globally since the input string is a strong key. This would make switching tabs super fast.
I'm not sure, but if the math extract/inject wrapper around DOMPurify is for MathML, I believe there's an option to persist MathML.
:+1:
1


developit  11 hours ago
I think the caching thing might be your best bet. It you move the entire flow of processing comments down to sanitized HTML off into a cached function, you'll see at least a 5% speedup.

cristian  11 hours ago
Top notch insights :+1: you're making me getting out of bed and look at it as well :joy: maybe
:smile:
1


developit  11 hours ago
and then for the useLayoutEffect thing, a debounced hook should work there:
import { useLayoutEffect as uLE } from 'preact/hooks';
const queue = [];
function useLayoutEffect(fn, deps) {
  uLE(() => {
    if (queue.indexOf(fn)===-1 && queue.push(fn)===1) {
      setTimeout(process, 0);  // maybe rAF instead.
    }
  }, deps);
}
function process() {
  queue.forEach(f => f());
  queue.length = 0;
}

developit  11 hours ago
or, instead of using render() to re-render the Preact tree in your updateReactComponent() method, create the tree using render() in a mount callback (on the angular side, I forget the name), but wrap everything in a root component (basically like your ServiceContext.Provider), then use setState()  on that root component to trigger rendering.  Doing this will place all of the different preact roots into the same update queue, and layout effects are flushed at the end of that shared queue

developit  10 hours ago
alright I built a whole demo for this lol

developit  10 hours ago
https://jsfiddle.net/developit/u28yswx7/
jsfiddle.netjsfiddle.net
Edit fiddle - JSFiddle
Test your JavaScript, CSS, HTML or CoffeeScript online with JSFiddle code editor.

developit  10 hours ago
it's using custom elements instead of angular, but it proves that my suggestion of using setState() actually doesn't help.

developit  10 hours ago
image.png 
image.png



developit  10 hours ago
as it turns out, our queuing technique for useEffect()  is actually probably what you want here. Even with multiple top-level render() calls, the fact that useEffect() is queued globally means they all run at the end.

robertknight  4 minutes ago
Oh wow Jason. Thank-you for all the time you spent looking into this and the insights!
Regarding useEffect vs useLayoutEffect, it is important that the effect runs before paint in this case to prevent a “flash of incorrectly styled content”, as the effect ends up setting (or unsetting) the max-heightof an element.
In the not-too-distant future, when the Angular => Preact conversion is a bit further along, we should have just one render root for the entire contents of the tab. It sounds like that will resolve the issue with useLayoutEffect?
```

</details>

In summary:

- The `useLayoutEffect` hook in `Excerpt` is triggering a series of forced layouts due to accessing `clientHeight`. This is interspersed with rendering because each Angular-Preact wrapper component is a separate "root". When all the tab contents become part of a single Preact component, this should be resolved
- Markdown/math rendering does some expensive context creation in DOMPurify. We might be able to optimize this
- `@sentry/browser` appears to be adding some inefficient debouncing to `addEventListener`

**Comment by robertknight** [comment]
We have solved the issue mentioned in the first bullet point because:

> When all the tab contents become part of a single Preact component, this should be resolved

The other two bullet points might still apply. However in my local testing, I cannot reproduce the user-facing issue any more, so I'm going to close this.
